core/src/connection_reuse.rs

```diff
@@ -40,193 +40,543 @@
 use parking_lot::Mutex;
-use std::io::{self, Error as IoError};
-use std::sync::Arc;
+use std::collections::{HashMap, hash_map::Entry};
+use std::io::{Error as IoError, ErrorKind as IoErrorKind, Read, Write};
+use std::ops::{Deref, DerefMut};
+use std::sync::{atomic::AtomicUsize, atomic::AtomicBool, atomic::Ordering, Arc};
 use tokio_io::{AsyncRead, AsyncWrite};
 use transport::{MuxedTransport, Transport, UpgradedNode};
 use upgrade::ConnectionUpgrade;
 
+use std::hash::{BuildHasherDefault, Hasher};
+
+#[derive(Default)]
+struct UsizeHasher {
+    state: usize
+}
+
+impl Hasher for UsizeHasher {
+    fn finish(&self) -> u64 {
+        self.state as u64
+    }
+    fn write(&mut self, _i: &[u8]) {
+        unreachable!()
+    }
+    fn write_usize(&mut self, i: usize) {
+        self.state = i;
+    }
+}
+
+type UsizeHasherBuilder = BuildHasherDefault<UsizeHasher>;
+
+
+
 /// Allows reusing the same muxed connection multiple times.
 ///
 /// Can be created from an `UpgradedNode` through the `From` trait.
 ///
-/// Implements the `Transport` trait.
+/// Implements the `Transport` trait
 #[derive(Clone)]
 pub struct ConnectionReuse<T, C>
 where
     T: Transport,
-    T::Output: AsyncRead + AsyncWrite,
+    C: ConnectionUpgrade<T::Output, T::MultiaddrFuture> + Clone,
+    C::Output: StreamMuxer + Clone,
+{
+    /// Struct shared between most of the `ConnectionReuse` infrastructure.
+    manager: Arc<ConnectionsManager<T, C>>,
+}
+
+/// Keeps an internal counter, for every new number issued
+/// will increase its internal state.
+#[derive(Default)]
+struct Counter {
+    inner: AtomicUsize,
+}
+
+impl Counter {
+    /// Returns the next counter, increasing the internal state
+    pub fn next(&self) -> usize {
+        self.inner.fetch_add(1, Ordering::Relaxed)
+    }
+}
+
+enum PeerState<M>
+where
+    M: StreamMuxer + Clone,
+{
+    /// Connection is active and can be used to open substreams.
+    Active {
+        /// The muxer to open new substreams.
+        muxer: M,
+        /// Id of the listener that created this connection, or `None` if it was opened by a
+        /// dialer.
+        listener_id: Option<usize>,
+        /// whether this connection was closed
+        closed: Arc<AtomicBool>
+    },
+
+    /// Connection is pending.
+    // TODO: stronger Future type
+    Pending {
+        /// Future that produces the muxer.
+        future: Box<Future<Item = (M, Multiaddr), Error = IoError>>,
+        /// All the tasks to notify when `future` resolves.
+        notify: HashMap<usize, task::Task, UsizeHasherBuilder>,
+    },
+
+    /// An earlier connection attempt errored.
+    Errored(IoError),
+}
+
+impl<M> PeerState<M>
+where
+    M: StreamMuxer + Clone
+{
+    /// resets the connections if the given PeerState is active and has
+    /// substreams. Consume the PeerState while doing so.
+    fn reset_conn(self) {
+        if let PeerState::Active { closed, .. } = self {
+            closed.store(false, Ordering::Release);
```

Comment:
I'm not a fan of having `Ordering::Release` without any context, as the ordering to use strongly depends on the context.
Personally I'd just copy paste these three lines of code at the two places where we use them.

Reply:
I generally agree with you regarding copying three lines around. The reason I went for this being a constraint within a function rather than copied is that this isn't yet fully resetting the connection yet. So, I can see us quickly expanding what should happen here (by first adding some debugging statements, later doing some actual signaling/cleaning up of the related connections) and if you then had to do that at many code location, you easily miss one when doing changes and most reviewer(s) won't notice. If you insist, though, I don't have a strong opinion on this and can do a copy instead.

Regarding the `Ordering:Release`, I am not sure what else we would want to do here/at the specific locations. @twittner pointed out correctly, [that `Ordering:Relaxed` might simply never gets propagated](https://github.com/libp2p/rust-libp2p/pull/387#discussion_r208512061). I don't think that we need to expect a strong guarantee of ordering in this case, as it doesn't actually hurt if you read the remaining bytes of an already closed connection. The only important things here, to me is, that it will _eventually_ fail and not dangle around for too long.